bpf: remove MapInfo, DumpParser and MapKey/Value DeepCopy

ti-mo opened the pull request:
Please review per commit.

High-level overview of this change:

- Replace MapInfo with MapSpec and accessors on Map
- Reduce MapKey and MapValue interfaces to just fmt.Stringer and New(), making it significantly easier to implement new key/value types
- Make OpenMap more usable by taking MapKey and MapValue arguments
- Remove k8s deepcopy machinery from all MapKey and MapValue types

```release-note
bpf: remove MapInfo, DumpParser and MapKey/Value DeepCopy
```

ti-mo committed:
bpf: replace GetMapInfo with ebpf.Map in objCheck

The ebpf.Map contained in a bpf.Map already contains the necessary map info for
running the comparison in objCheck.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: replace MapInfo with ebpf.MapSpec and MapType with ebpf.MapType

This commit removes pkg/bpf.MapType and inlines its allowsPreallocation
and requiresPreallocation methods into GetPreallocateMapFlags. Also
removes its MapType* enum and replaces all of its uses by constants
exported from ebpf-go.

GetPreallocateMapFlags and NewMap now take an ebpf.MapType instead of a
pkg/bpf.MapType. NewMap and friends no longer take key/value sizes.

Removed pkg/bpf.MapInfo in favor of Map.spec field containing an *ebpf.MapSpec.
Getter methods like Map.Type(), Map.MaxEntries() etc. replace direct MapInfo
field accesses.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove MapKey.GetKeyPtr() and MapValue.GetValuePtr()

Currently, the MapKey and MapValue interfaces are a little bloated for what
they're supposed to accomplish: identifying types to be used as map keys and
values.

The MapKey and MapValue interfaces provide useful methods, but also contain
a few that are either not strictly necessary, sidestep Go's runtime safety,
actively hurt performance, or all three. This series of patches aims to
drive down the burden of implementing and maintaining k/v types by removing
some of these methods, starting by GetKeyPtr() and GetValuePtr().

The legacy (now removed) implementations of map operations (update, lookup,
delete) had underlying primitives that operated exclusively on unsafe.Pointers.
The burden of returning those pointers was put on the MapKey and MapValue
types themselves by requiring the implementation of methods that returned them.

With the move to ebpf-go-based map operations, this is no longer necessary, as
standard Go pointers are now supported, and values are marshaled to/from []byte
under the hood. The new implementation also checks if the length of the input
type corresponds to the key or value type configured on the map, which was not
previously possible with the opaque unsafe.Pointer approach. This change surfaced
cases where structs used as map k/v were lacking the appropriate padding.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove UnpinMapIfExists(), use Map.Unpin()

This was only used in tests, not sure why it was useful. Works on closed Maps
regardless, and was an glorified (and expensive) version of os.RemoveAll().

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf, maps: remove DumpParser and ConvertKeyValue

ebpf-go already has the capability to unmarshal map keys and values into any
given pointer type. DumpParser was only ever nil or ConvertKeyValue, so cut it
out entirely and let the library unmarshal directly into MapKey and MapValue.

This means we no longer need pkg/bpf/binary, remove it.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove MapKey.NewValue()

This ended up only being used by lbmap, which has its own ServiceKey and
ServiceValue interfaces. Added ServiceValue.New() instead.

Remove NewValue() from all types implementing MapKey.

Fix a few linter errors regarding unused struct padding.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: remove MapKey/Value footgun from OpenMap

It was documented in a few places that OpenMap requires the caller to manually
set Map.MapKey and MapValue. Make OpenMap take those values as parameters
instead.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: unexport Map.MapKey and Map.MapValue

Since these are really for internal use only (returning the correct types
from Lookup and Dump*), unexport the variable.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf, maps: remove k8s DeepCopy generators from all bpf.MapKey/Values

Since the DeepCopy semantic for map k/v types was basically unused, remove it
and all the code generation needed to support it. Implementing new k/v types
is now much more straightforward.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
maps, daemon: remove use of MapKey/Value DeepCopies outside of package bpf

To reduce the bpf.MapKey and bpf.MapValue interface footprints, wind down usage
of DeepCopy methods outside of package bpf. A future commit will replace this
method with New(), which is currently what it's primarily abused for (in
DumpWithCallback and friends).

Map keys and values cannot contain pointer types, so the presence of deep copy
methods is questionable. Copying these types by value should be sufficient to
achieve 'deep' copying behaviour.

The changes in this patch aim to be no-ops:
- The ToHost() methods implemented on lbmap.BackendValue, AffinityMatchKey and
  ServiceValue all copy their receivers by value and already return new copies.
- The callback in policymap dereferences the pointer (and DumpCallbacks already
  receive pointers to unshared copies), so an additional copy is unnecessary.
- The lxcmap callback was changed to dereference the pointers received in the
  callback and return a map of EndpointInfos instead of pointers.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf, maps: rename DeepCopyMapKey/Value() interface methods to New()

Since the deep copying semantic was removed in a prior commit, rename the
interface methods and all their implementations to New().

Signed-off-by: Timo Beckers <timo@isovalent.com>

lmb commented on pkg/bpf/map.go:
```diff
@@ -10,33 +10,8 @@ import (
-// MapType is an enumeration for valid BPF map types
-type MapType int
-
 // This enumeration must be in sync with enum bpf_map_type in <linux/bpf.h>
```
Nit: out of date.

lmb commented on pkg/bpf/bpf.go:
```diff
@@ -38,12 +40,15 @@ func DisableMapPreAllocation() {
-		return 0
+	// Support disabling preallocation for these map types.
+	case ebpf.Hash, ebpf.PerCPUHash, ebpf.HashOfMaps:
+		return atomic.LoadUint32(&preAllocateMapSetting)
```
It's kind of weird that this returns preAllocateMapSetting directly. This only works because BPF_F_NO_PREALLOC is 1 like the constant in DisableMapPreAllocation.

tommyp1ckles replied:
nit: seems like a good use for atomic.Bool?

ti-mo replied:
The fact that this is a global to begin with is a red flag. We'll need to sort this out when map creation in `pkg/bpf` takes MapSpec. Currently, the global is only manipulated from tests.

With MapSpec in the picture, we'd only set the default map flags based on option.Config or leave it untouched if the caller populates the flags explicitly in the spec (such as from tests, etc.).

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -81,9 +67,11 @@ type cacheEntry struct {
-	m         *ebpf.Map
-	innerSpec *ebpf.MapSpec
+	m    *ebpf.Map
+	spec *ebpf.MapSpec
```
Nit: `// Nil after the map has been created`?

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -1142,17 +1166,17 @@ func (m *Map) resolveErrors(ctx context.Context) error {
 // Returns true if the map was upgraded.
-func (m *Map) CheckAndUpgrade(desired *MapInfo) bool {
-	desired.Flags |= GetPreAllocateMapFlags(desired.MapType)
+func (m *Map) CheckAndUpgrade(desired *Map) bool {
```
Would it make sense to pass `*ebpf.MapSpec` here?

ti-mo replied:
MapSpec is no longer available after a Map had OpenOrCreate() called on it, or when it was opened from a pin. CheckAndUpgrade will be removed in 1.15. Calling Recreate() should be the responsibility of individual subsystems, as only they possess the knowledge of whether or not recreating a map causes data loss, and if so, it needs to take care of any migration, logging, etc. This is was one responsibility too many that was delegated to pkg/bpf.

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -570,33 +560,27 @@ type MapValidator func(path string) (bool, error)
-			return err
-		}
+	for i.Next(mk, mv) {
+		cb(mk, mv)
```
The function doc makes it sound like `cb` is responsible for copying mk and mv if it wants to hold onto them, but here were creating new copies for each iteration. Is the doc incorrect or should this not create copies?

ti-mo replied:
I'll fix the doc, we're changing this (and removing Dump* altogether) in the next cycle.

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -627,14 +611,23 @@ func (m *Map) DumpWithCallbackIfExists(cb DumpCallback) error {
+	}
+
 	var (
-		prevKey    = make([]byte, m.KeySize())
```
Using `[]byte` for prevKey / nextKey is a lot cheaper even when using the library, since it avoids marshaling. See the gymnastics that `ebpf.MapIterator` does.

ti-mo replied:
Since we need to pass all keys and values to cb() in an unmarshaled state, what would this gain us?

lmb replied:
Just wanted to raise this as an FYI, all good if you've done the due diligence.

lmb commented on pkg/bpf/map_linux.go:
```diff
@@ -980,44 +962,26 @@ func (m *Map) DeleteAll() error {
-			log.WithError(err2).Warningf("Unable to correlate iteration key %v with cache entry. Inconsistent cache.", key)
-		}
+	for i.Next(mk, mv) {
+		err := m.m.Delete(mk)
```
I guess this will hit the "concurrent delete from hash map resets iteration" case, no? Is there a test that checks we actually delete all elements?

ti-mo replied:
True, but in this case, that's intended. Each GetNext call will return the first key, which we then immediately delete, until all records are gone. The process continues until GetNext is called on a map that has no keys, which could never happen if the insertion rate is high enough. This method is not advertised to deal with the termination problem, though, and we're not going to change it in this series.

ti-mo replied:
Looking at the code again, I noticed I put unnecessary `m.{key,value}.New()` etc. in the loop, I've removed those.

lmb commented on pkg/maps/ctmap/ctmap.go:
```diff
@@ -709,7 +709,7 @@ func DeleteIfUpgradeNeeded(e CtEndpoint) {
 		scopedLog := log.WithField(logfields.Path, path)
 
-		oldMap, err := bpf.OpenMap(path)
+		oldMap, err := bpf.OpenMap(path, nil, nil)
```
Why is `nil, nil` okay here? Maybe worth documenting on OpenMap.

ti-mo replied:
I've added a comment to the call site, doesn't make much sense to call this out in the method doc IMO.

ti-mo committed:
bpf: add benchmark for Map.Lookup()

Signed-off-by: Timo Beckers <timo@isovalent.com>

tommyp1ckles commented on pkg/maps/bwmap/bwmap.go:
```diff
@@ -43,8 +41,7 @@ type EdtInfo struct {
 
-func (v *EdtInfo) GetValuePtr() unsafe.Pointer { return unsafe.Pointer(v) }
-func (v *EdtInfo) String() string              { return fmt.Sprintf("%d", int(v.Bps)) }
+func (v *EdtInfo) String() string { return fmt.Sprintf("%d", int(v.Bps)) }
```
I know this is inherited code, but do we need to cast down to a signed int, seems like a recipe for trouble.

ti-mo replied:
I'm trying to avoid altering behaviours like these in this PR. Agree overall, this was likely copypasta'd, but wouldn't cause anything worse than a cosmetical bug.

tommyp1ckles commented on pkg/bpf/bpf.go:
```diff
@@ -38,12 +40,15 @@ func DisableMapPreAllocation() {
-	case t.requiresPreallocation():
-		return 0
+	// Support disabling preallocation for these map types.
+	case ebpf.Hash, ebpf.PerCPUHash, ebpf.HashOfMaps:
```
question - we don't support this for trie type anymore?

ti-mo replied:
Could you clarify the question? Tries don't support preallocation.

tommyp1ckles replied:
Think I was referencing this deleted code: https://github.com/cilium/cilium/blob/main/pkg/bpf/map.go#L106

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -570,33 +560,27 @@ type MapValidator func(path string) (bool, error)
-	value := make([]byte, m.ValueSize())
-
+	// Don't need deep copies here, only fresh pointers.
 	mk := m.MapKey.DeepCopyMapKey()
```
why don't we just call `New()` here?

ti-mo replied:
That's changed in a subsequent commit. :) Essentially, DeepCopyMap* was repurposed to mean New().

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -570,33 +560,27 @@ type MapValidator func(path string) (bool, error)
-		if cb != nil {
-			cb(mk, mv)
-		}
+		mk = m.MapKey.DeepCopyMapKey()
```
why do we need to make copies here (following the callback invocation)?

ti-mo replied:
Good question! Handing out copies to callbacks is what the existing implementation does (even though docstrings claim the opposite), and we don't want to change that behaviour in this PR.

In the next cycle, we'll remove the `Dump*` concept altogether in favor of MapIterator, which has semantics closer to those of Go's `for` (so, unmarshal the next key and value into a caller-provided pointer argument). This will remove the ambiguity (and improve performance!), but requires going through the existing codebase to audit and potentially fix up all spots where callback arguments are stuffed directly into slices/maps etc. without copying.

tommyp1ckles commented on pkg/bpf/map_linux_test.go:
```diff
@@ -774,3 +776,39 @@ func (s *BPFPrivilegedTestSuite) TestCreateUnpinned(c *C) {
 	c.Assert(got, checker.DeepEquals, v)
 }
+
+func BenchmarkMapLookup(b *testing.B) {
```
Q: What's the context for this benchmark, are we proving that the ebpf-go has better overhead over the old Cilium bpf.Map?

ti-mo replied:
Yes, there is a rather large performance regression due to getting rid of unsafe.Pointer in favor of reflect-based unmarshaling in most of the maps. We're currently addressing this in ebpf-go by adding a bytes.Buffer pool on the key marshaling path in the short term and investigating in which cases we can pass unsafe.Pointers to the kernel internally. The latter will require a few more cycles and won't make it in time for the 1.14 freeze.

ti-mo replied:
ebpf-go PR for this is up: https://github.com/cilium/ebpf/pull/1053

ti-mo replied:
Upstream PR merged and pulled into vendor in this PR.

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -124,54 +112,99 @@ type Map struct {
-	mapValue MapValue, valueSize, maxEntries int, flags uint32, dumpParser DumpParser) *Map {
+func (m *Map) Type() ebpf.MapType {
+	if m.m != nil {
+		return m.m.Type()
```
question: it will follow this code path if the map has been created, or use the spec struct otherwise?

ti-mo replied:
Yep, that's the idea!

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -286,7 +319,7 @@ func (m *Map) updatePressureMetric() {
 	}
 
-	pvalue := float64(len(m.cache)) / float64(m.MaxEntries)
+	pvalue := float64(len(m.cache)) / float64(m.MaxEntries())
```
It looks like the `MaxEntries()` func can return 0 in some cases, not sure what happens here in that situation.

ti-mo replied:
This could occur before this change as well, I don't think we ever validated this piece of input. Seeing as `updatePressureMetric()` is only ever invoked after calls to `Map.open()`, which is guaranteed to fail with maxentries 0, I don't think this is an issue.

tommyp1ckles commented on pkg/bpf/map_linux.go:
```diff
@@ -124,54 +112,99 @@ type Map struct {
+func NewMap(name string, mapType ebpf.MapType, mapKey MapKey, mapValue MapValue,
+	maxEntries int, flags uint32, dumpParser DumpParser) *Map {
+
+	keySize := reflect.TypeOf(mapKey).Elem().Size()
```
note: this panics if mapKey is not a pointer

ti-mo replied:
You've flagged this in a previous PR as well. Currently, all code assumes that MapKey and MapValue are implemented on pointer receivers. Panicking isn't a friendly way of enforcing that, but this also isn't too high up on the list of things to tackle. We can make this friendlier when NewMap() is changed to take a MapSpec, but eventually, bpf.Map should no longer hold MapKey and MapValue, so this would be somewhat wasted effort.

asauber reviewed (approved):
CLI-related changes LGTM

ti-mo committed:
vendor: bump cilium/ebpf dependency to unreleased master

A subsequent commit will get rid of doing map operations using unsafe.Pointer
in favor of passing plain, safe Go pointers. ebpf-go was not re-using
bytes.Buffers in the key marshaling path, leading to a rather severe performance
hit. This shortcoming has been addressed on cilium/ebpf@master, but there is
no release yet.

Signed-off-by: Timo Beckers <timo@isovalent.com>

jrajahalme reviewed (approved):
LGTM for my "CODEOWNERS"
